### PR TITLE
#12739: Added allCategoriesInfo parameter for retrieve all categories info in Content REST API

### PR DESCRIPTION
**Pull request by dsolistorres**
This PR adds the `allCategoriesInfo` parameter to Content REST API:
```
/api/content/id/{ContentIdentifier}/allCategoriesInfo/1
```
When the parameter is used, the full categories information is returned for the content. When the parameter is not used (the default behavior), only category names are included in the result.
For example, this info is generated when the parameter is used:
```
{
  "contentlets": [
    {
    ...
    "topic": [
        {
          "inode": "ede13663-ff06-4d01-ab99-a8976e995010",
          "keywords": "",
          "sortOrder": 0,
          "description": null,
          "categoryName": "Research",
          "key": "research"
        },
        {
          "inode": "bba0fcba-eb79-46dc-ae36-fe2f3d9cf1e7",
          "keywords": "",
          "sortOrder": 0,
          "description": null,
          "categoryName": "Retirement",
          "key": "retirement"
        }
      ]
    ...
    }
  ]
}
```

### CODE REVIEW

File: dotCMS/src/integration-test/java/com/dotmarketing/portlets/contentlet/util/ContentletUtilTest.java [hunk 1]
```diff
@@ -377,43 +270,32 @@ public void test_getContentPrintableMap_WhenContentTypeIsFileAsset_PathIsAddedTo
...
+        assertEquals(categories.length, categoryList.size());
+        for (final Category category : categories) {
+            Optional<?> matchElement = categoryList.stream()
+                    .filter(map -> {
+                        final Map<?,?> cMap = (Map<?,?>) map;
+                        return cMap.containsKey("key")
+                                && cMap.get("key").equals(category.getKey());
+                    }).findFirst();
+            assertTrue(matchElement.isPresent());
+            Map<?,?> categoryMap = (Map<?, ?>) matchElement.get();
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/portlets/contentlet/util/ContentletUtilTest.java]: Codacy Issue found: [Local variable 'categoryMap' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4885353)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentResourceTest.java [hunk 2]
```diff
@@ -1216,6 +1373,15 @@ private User createUser() throws DotSecurityException, DotDataException {
         return newUser;
     }
 
+    private CategoryDataGen createCategory(final String categoryName, final int sortOrder) {
+        SimpleDateFormat simpleDateFormat = new SimpleDateFormat("yyyyMMdd-HHmmss");
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentResourceTest.java]: Codacy Issue found: [Local variable 'simpleDateFormat' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4885353)

File: dotCMS/src/integration-test/java/com/dotmarketing/portlets/contentlet/util/ContentletUtilTest.java [hunk 3]
```diff
@@ -377,43 +270,32 @@ public void test_getContentPrintableMap_WhenContentTypeIsFileAsset_PathIsAddedTo
...
-            }
+    private void verifyCategoriesForField(
+            final Map<String, Object> contentPrintableMap,
+            final String fieldName, final Category ...categories) {
+
+        final List<?> categoryList = (List<?>) contentPrintableMap.get(fieldName);
+        assertNotNull(categoryList);
+        assertEquals(categories.length, categoryList.size());
+        for (final Category category : categories) {
+            Optional<?> matchElement = categoryList.stream()
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/portlets/contentlet/util/ContentletUtilTest.java]: Codacy Issue found: [Local variable 'matchElement' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4885353)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentResourceTest.java [hunk 4]
```diff
@@ -921,6 +1050,34 @@ private void validateChildrenXML(final Map<String, Contentlet> contentletMap,
...
     }
 
+    private void validateCategoryXML(
+            final String resultXml, final String categoryElement, final Category... categories)
+            throws XPathExpressionException {
+
+        final InputSource xmlSource = new InputSource(new StringReader(
+                resultXml.replaceAll("\\n", "")));
+
+        XPath xPath = XPathFactory.newInstance().newXPath();
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentResourceTest.java]: Codacy Issue found: [Local variable 'xPath' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4885353)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentResourceTest.java [hunk 5]
```diff
@@ -921,6 +1050,34 @@ private void validateChildrenXML(final Map<String, Contentlet> contentletMap,
...
+
+        NodeList itemList = (NodeList) xPath.evaluate("//customTopic/item",
+                xmlSource, XPathConstants.NODESET);
+
+        assertEquals(categories.length, itemList.getLength());
+        for (int i = 0; i < itemList.getLength(); i++) {
+            final Element resultCategory = (Element) itemList.item(i);
+            final String resultCategoryKey = (String) xPath.evaluate("key",
+                    resultCategory, XPathConstants.STRING);
+            Optional<Category> matchCategory = Arrays.stream(categories)
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentResourceTest.java]: Codacy Issue found: [Local variable 'matchCategory' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4885353)

File: dotCMS/src/integration-test/java/com/dotmarketing/portlets/contentlet/util/ContentletUtilTest.java [hunk 6]
```diff
@@ -64,251 +70,138 @@ public static void prepare () throws Exception {
...
+            assertTrue(contentPrintableMap.containsKey(CATEGORY_NAME_CONTENT));
+            assertTrue(contentPrintableMap.containsKey(CATEGORY_NAME_CONTENT_BELTS));
+            if (categoryInfoType.equals(ALL_CATEGORIES_INFO)) {
+                verifyCategoriesForField(contentPrintableMap, CATEGORY_NAME_CONTENT,
+                        popularCategory);
+                verifyCategoriesForField(contentPrintableMap, CATEGORY_NAME_CONTENT_BELTS,
+                        homeCategory, flightsCategory);
+            } else {
+                assertEquals( popularCategory.getCategoryName(), contentPrintableMap.get( CATEGORY_NAME_CONTENT ) );
+                String [] resultCategories = contentPrintableMap.get(CATEGORY_NAME_CONTENT_BELTS)
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/portlets/contentlet/util/ContentletUtilTest.java]: Codacy Issue found: [Local variable 'resultCategories' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4885353)

File: dotCMS/src/main/java/com/dotmarketing/portlets/contentlet/util/ContentletUtil.java [hunk 7]
```diff
@@ -118,16 +117,39 @@ public static String sanitizeFileName(String fileName){
...
+	 * the categories associated to the content (key, name, description),{@code "false"}
+	 * to return only categories names.
+	 *
+	 * @return Contentlet with the values in place.
+	 *
+	 * @throws DotDataException
+	 * @throws IOException
+	 */
+	public static Map<String, Object> getContentPrintableMap(
+			final User user, Contentlet contentlet, final boolean allCategoriesInfo)
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/contentlet/util/ContentletUtil.java]: Codacy Issue found: [Avoid reassigning parameters such as 'contentlet'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4885353)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentResourceTest.java [hunk 8]
```diff
@@ -921,6 +1050,34 @@ private void validateChildrenXML(final Map<String, Contentlet> contentletMap,
...
+                    .filter(category -> category.getKey().equals(resultCategoryKey))
+                    .findFirst();
+            assertTrue(matchCategory.isPresent());
+            Category expectedCategory = matchCategory.get();
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentResourceTest.java]: Codacy Issue found: [Local variable 'expectedCategory' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4885353)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentResourceTest.java [hunk 9]
```diff
@@ -921,6 +1050,34 @@ private void validateChildrenXML(final Map<String, Contentlet> contentletMap,
         }
     }
 
+    private void validateCategoryXML(
+            final String resultXml, final String categoryElement, final Category... categories)
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentResourceTest.java]: Codacy Issue found: [Avoid unused method parameters such as 'categoryElement'.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4885353)

File: dotCMS/src/integration-test/java/com/dotmarketing/portlets/contentlet/util/ContentletUtilTest.java [hunk 10]
```diff
@@ -64,251 +70,138 @@ public static void prepare () throws Exception {
...
+    }
 
     /**
      * https://github.com/dotCMS/core/issues/11751
      */
     @Test
-    public void validateContentPrintableMapMethodReturnProperCategories() throws Exception{
+    @UseDataProvider("categoryTestCases")
+    public void validateContentPrintableMapMethodReturnProperCategories(
+            final String categoryInfoType) throws Exception {
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/portlets/contentlet/util/ContentletUtilTest.java]: Codacy Issue found: [A method/constructor should not explicitly throw java.lang.Exception](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4885353)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentResourceTest.java [hunk 11]
```diff
@@ -921,6 +1050,34 @@ private void validateChildrenXML(final Map<String, Contentlet> contentletMap,
...
+            throws XPathExpressionException {
+
+        final InputSource xmlSource = new InputSource(new StringReader(
+                resultXml.replaceAll("\\n", "")));
+
+        XPath xPath = XPathFactory.newInstance().newXPath();
+
+        NodeList itemList = (NodeList) xPath.evaluate("//customTopic/item",
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentResourceTest.java]: Codacy Issue found: [Local variable 'itemList' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4885353)

File: dotCMS/src/integration-test/java/com/dotmarketing/portlets/contentlet/util/ContentletUtilTest.java [hunk 12]
```diff
@@ -377,43 +273,34 @@ public void test_getContentPrintableMap_WhenContentTypeIsFileAsset_PathIsAddedTo
...
-            }
+    private void verifyCategoriesForField(
+            final Map<String, Object> contentPrintableMap,
+            final String fieldName, final Category ...categories) {
+
+        final List<?> categoryList = (List<?>) contentPrintableMap.get(fieldName);
+        assertNotNull(categoryList);
+        assertEquals(categories.length, categoryList.size());
+        for (final Category category : categories) {
+            Optional<?> matchElement = categoryList.stream()
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/portlets/contentlet/util/ContentletUtilTest.java]: Codacy Issue found: [Local variable 'matchElement' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4885353)

File: dotCMS/src/integration-test/java/com/dotmarketing/portlets/contentlet/util/ContentletUtilTest.java [hunk 13]
```diff
@@ -377,43 +273,34 @@ public void test_getContentPrintableMap_WhenContentTypeIsFileAsset_PathIsAddedTo
...
+                    .filter(map -> {
+                        final Map<?,?> cMap = (Map<?,?>) map;
+                        return cMap.containsKey("key")
+                                && cMap.get("key").equals(category.getKey());
+                    }).findFirst();
+            assertTrue(matchElement.isPresent());
+            Map<?,?> categoryMap = (Map<?, ?>) matchElement.get();
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/portlets/contentlet/util/ContentletUtilTest.java]: Codacy Issue found: [Local variable 'categoryMap' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4885353)

File: dotCMS/src/main/java/com/dotmarketing/portlets/contentlet/util/ContentletUtil.java [hunk 14]
```diff
@@ -142,33 +164,37 @@ public static String sanitizeFileName(String fileName){
...
+							if (allCategoriesInfo) {
+								categoriesValue = childCategories.stream()
+										.map(Category::getMap)
+										.collect(Collectors.toList());
+							} else {
+								categoriesValue = childCategories.stream()
+										.map(Category::getCategoryName)
+										.collect(Collectors.joining(", "));
+							}
+							m.put(f.variable(), categoriesValue);
```
Comment (wezell) [inline dotCMS/src/main/java/com/dotmarketing/portlets/contentlet/util/ContentletUtil.java]: This is a great change and the code to include the categories is in the right place.

**Review by wezell — APPROVED** [review]

**Pull request merged**